src/efftester.ml

```diff
@@ -1072,209 +1152,314 @@ let shrinker term = wrapper termShrinker term
-    else failwith ("tcheck: If with non-Boolean condition")
+      then (
+        let mtyp, meff = tcheck env m in
+        let ntyp, neff = tcheck env n in
+        match unify mtyp ntyp with
+        | Sol sub ->
+          if types_compat (subst sub mtyp) t
+             (* we obtain annot by instantiating inferred type *)
+          then
+            if types_compat (subst sub ntyp) t
+               (* we obtain annot by instantiating inferred type *)
+            then
+              if eff_leq meff e && eff_leq neff e
+              then (
+                let e' = eff_join beff (eff_join meff neff) in
+                (t, e'))
+              else failwith "tcheck: If's branch effects disagree with annotation"
+            else
+              failwith
+                ("tcheck: If's else branch type disagrees with annotation: "
+                ^ "  term is "
+                ^ toOCaml ~typeannot:false term
+                ^ "  ntyp is "
+                ^ typetostr ~effannot:true ntyp
+                ^ "  (subst sub ntyp) is "
+                ^ typetostr ~effannot:true (subst sub ntyp)
+                ^ "  t is "
+                ^ typetostr ~effannot:true t)
+          else
+            failwith
+              ("tcheck: If's then branch type disagrees with annotation: "
+              ^ "  term is "
+              ^ toOCaml ~typeannot:false term
+              ^ "  mtyp is "
+              ^ typetostr ~effannot:true mtyp
+              ^ "  (subst sub mtyp) is "
+              ^ typetostr ~effannot:true (subst sub mtyp)
+              ^ "  t is "
+              ^ typetostr ~effannot:true t)
+        | No_sol ->
+          failwith
+            ("tcheck: If's branch types do not unify:  "
+            ^ "  term is "
+            ^ toOCaml ~typeannot:false term
+            ^ "  mtyp is "
+            ^ typetostr ~effannot:true mtyp
+            ^ "  ntyp is "
+            ^ typetostr ~effannot:true ntyp))
+      else failwith "tcheck: If's condition effect disagrees with annotation"
+    else failwith "tcheck: If with non-Boolean condition"
+;;
 
 let print_wrap t =
-  Let ("i",Int,t,
-       App (Unit,
-	    Variable (Fun (Int,(true,false),Unit),
-		      "print_int"),
-	    Int,
-	    Variable (Int,"i"),
-	     (true,false)),
-       Unit, (true,false))
+  Let
+    ( "i",
+      Int,
+      t,
+      App
+        ( Unit,
+          Variable (Fun (Int, (true, false), Unit), "print_int"),
+          Int,
+          Variable (Int, "i"),
+          (true, false) ),
+      Unit,
+      (true, false) )
+;;
 
 let term_gen =
-  make ~print:(Print.option (toOCaml ~typeannot:false))
-(*       ~shrink:shrinker *)
-    (listPermuteTermGenRecWrapper initTriEnv Int (true,false)) (* Initial goal here! *)
+  make
+    ~print:(Print.option (toOCaml ~typeannot:false))
+    (*       ~shrink:shrinker *)
+    (listPermuteTermGenRecWrapper initTriEnv Int (true, false))
+;;
+
+(* Initial goal here! *)
```

Comment:
I suggest putting the comment within the last parenthesis.
The point is to highlight where the goal is: "generate an int program which can have an effect".
If the comment follows two lines after, that meaning is lost.